### PR TITLE
PT-161183355 Use tx gas to check minimal tx fee

### PR DESCRIPTION
**Pull request by jur0**
[PT-161183355](https://www.pivotaltracker.com/n/projects/2124891/stories/161183355)

The tx gas is now used to compute the minimal tx fee (min_fee = gas * min_gas_price)

TODO:
- [x] add minimal tx fee check into micro block validation
- [x] python UATs
- [x] integration tests 
- [x] system tests

### CODE REVIEW

File: apps/aecore/src/aec_tx_pool.erl [hunk 1]
```diff
@@ -742,8 +742,9 @@ get_account(AccountKey, {account_trees, AccountsTrees}) ->
 get_account(AccountKey, {block_hash, BlockHash}) ->
     aec_chain:get_account_at_hash(AccountKey, BlockHash).
 
-check_minimum_fee(Tx,_Hash) ->
-    case aetx:fee(aetx_sign:tx(Tx)) >= aec_governance:minimum_tx_fee() of
+check_minimum_fee(Tx0, _Hash) ->
+    Tx = aetx_sign:tx(Tx0),
+    case aetx:fee(Tx) >= (aetx:gas(Tx) * aec_governance:minimum_gas_price()) of
```
Comment (hanssv) [inline apps/aecore/src/aec_tx_pool.erl]: This is not correct when Tx is `aect_create_tx` or `aect_call_tx` the fee should _not_ cover `Gas * GasPrice` in this case, it should only cover the basic transaction cost (base_gas + tx_size gas)?!

Comment (jur0) [reply apps/aecore/src/aec_tx_pool.erl]: @hanssv I updated it. The min fee of contract txs doesn't include the gas for contract execution.

**Comment by jur0** [comment]
There is one test failure, which is intermittent:
```
%%% aecore_txs_SUITE ==> txs_gc: FAILED
%%% aecore_txs_SUITE ==> {{badmatch,
     {2,
      [{signed_tx,
           {aetx,spend_tx,aec_spend_tx,81,
               {spend_tx,
                   {id,account,
                       <<206,167,173,228,112,201,249,157,157,78,64,8,128,168,
                         111,29,73,187,68,75,98,241,26,158,187,100,187,207,
                         235,115,254,243>>},
                   {id,account,
                       <<229,95,184,116,250,171,198,62,8,238,100,208,90,6,99,
                         187,138,81,179,95,20,151,222,65,118,201,24,5,153,49,
                         178,243>>},
                   1000,20000,10,7,<<>>}},
           [<<37,17,105,163,51,208,95,183,92,118,150,65,105,175,174,7,25,51,
              86,24,160,66,253,198,128,38,72,96,193,181,132,254,38,94,97,214,
              114,142,244,141,14,139,24,52,55,161,54,92,71,121,165,81,33,230,
              123,36,148,83,228,49,138,73,44,9>>]},
       {signed_tx,
           {aetx,spend_tx,aec_spend_tx,81,
               {spend_tx,
                   {id,account,
                       <<206,167,173,228,112,201,249,157,157,78,64,8,128,168,
                         111,29,73,187,68,75,98,241,26,158,187,100,187,207,
                         235,115,254,243>>},
                   {id,account,
                       <<130,154,92,204,168,237,223,55,198,213,211,128,17,134,
                         112,45,27,182,238,203,165,227,158,251,39,28,121,63,
                         53,32,68,51>>},
                   1000,20000,7,8,<<>>}},
           [<<144,252,103,88,188,162,188,192,109,2,175,205,51,43,41,74,41,
              92,96,57,62,187,90,164,1,78,203,105,107,29,127,112,165,34,
              28,136,97,134,65,231,158,32,144,180,164,4,251,116,135,207,
              147,4,217,240,90,10,84,2,251,91,106,221,63,8>>]}]}},
 [{aecore_txs_SUITE,txs_gc,1,
      [{file,
           "/home/builder/epoch/_build/test/lib/aecore/test/aecore_txs_SUITE.erl"},
       {line,147}]},
  {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1546}]},
  {test_server,run_test_case_eval1,6,[{file,"test_server.erl"},{line,1062}]},
  {test_server,run_test_case_eval,9,[{file,"test_server.erl"},{line,994}]}]}
```

File: apps/aetx/src/aetx.erl [hunk 2]
```diff
@@ -180,6 +181,19 @@ gas_price(#aetx{ type = Type, cb = CB, tx = Tx }) when
 gas_price(#aetx{}) ->
     undefined.
 
+-spec min_gas(Tx :: tx()) -> Gas :: non_neg_integer().
+min_gas(#aetx{ type = Type, size = Size }) when
+      Type =:= contract_create_tx;
+      Type =:= contract_call_tx;
+      Type =:= channel_force_progress_tx ->
+    aec_governance:tx_base_gas(Type) + Size * aec_governance:byte_gas();
+min_gas(#aetx{ type = Type, cb = CB, size = Size, tx = Tx }) when
```
Comment (hanssv) [inline apps/aetx/src/aetx.erl]: From here and down this function is identical to `gas/1`

**Comment by jur0** [comment]
@velzevur had a similar suggestion - rename `gas/1` to `gas_limit/1`. I'd probably keep it as it is for now. It'd be more intuitive to have `max_gas/1` or `gas_limit/1`, but I'd put it into a separate PR.

**Comment by jur0** [comment]
`gas/1` refactoring task tracked as https://www.pivotaltracker.com/n/projects/2124891/stories/161759909

**Pull request merged**